Sources/SourceControl/RepositoryManager.swift

```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
     /// Get a handle to a repository.
     ///
     /// This will initiate a clone of the repository automatically, if
-    /// necessary, and immediately return. The client can add observers to the
-    /// result in order to know when the repository is available.
-    public func lookup(repository: RepositorySpecifier) -> RepositoryHandle {
-        // Check to see if the repository has been provided.
-        if let handle = repositories[repository.url] {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
-            do {
-                let repo = try handle.open()
-                try repo.fetch()
-            } catch {
-                // FIXME: Better error handling.
-                print("Error while fetching the repo: \(handle)")
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (LookupResult) -> Void) {
+        DispatchQueue.global().async {
+            // Get the handle for this repository.
+            let handle = self.getHandle(repository: repository)
+            switch handle.status {
```

Comment:
This doesn't seem safe, the concurrent access to `handle.state` isn't protected by anything.

As a general rule, if you have some variable protected by a lock, it must *always* be accessed via that protection (google for why double checked locking isn't safe if you are curious). We need to find a pattern where the state is more protected.